### PR TITLE
`XspectraCalculation`+`XspectraParser`: Add Features for Restarts and Other Updates

### PR DESCRIPTION
**Pull request by PNOGillespie**
#### Adds features necessary for restarting XSpectra calculations:
* Reserves the save file name (`xanes.sav`) in the `codeinfo` object.
* The code now automatically includes `xanes.sav` in the `local_copy_list` if the parent calculation is an `XspectraCalculation` (taken from the parent folder), thus simplifying the setup for calculation restarts.
#### Other additions included in the PR:
* A new error code for the `XspectraParser` (code 315) to detect if the code has cleanly exited after reaching a set maximum time limit (including an accompanying test).
* A slight increase to the `output_parameters` collected: adds the polarisation vectors chosen for the calculation and whether the calculation was set to re-plot (i.e. `xonly_plot = .true./.false.` in the inputs).
* Support for including an optional `gamma_file` (`SinglefileData` format) required to run calculations using an external file for applying spectral broadening to the final spectrum (using `gamma_mode = "file"`).

### CODE REVIEW

File: src/aiida_quantumespresso/calculations/xspectra.py [hunk 1]
```diff
@@ -108,4 +127,21 @@ def prepare_for_submission(self, folder):
...
+            gamma_file_info = (gamma_file.uuid, gamma_file.filename, gamma_file.filename)
+            calcinfo.local_copy_list.append(gamma_file_info)
+
+        # Check if the parent folder is an XspectraCalculation type, if so we then copy the
+        # save file to enable calculation restarts and re-plots.
+        parent_folder = self.inputs.parent_folder
+        parent_calcs = parent_folder.get_incoming(node_class=orm.CalcJobNode).all()
+        parent_calc = parent_calcs[0].node
+        restart_flag = parent_calc.process_type == 'aiida.calculations:quantumespresso.xspectra'
+        if restart_flag:
```
Comment (sphuber) [inline src/aiida_quantumespresso/calculations/xspectra.py:142]: Here is a useful shortcut
```suggestion
        parent_folder = self.inputs.parent_folder
        parent_calc = parent_folder.creator

        if parent_calc.process_type == 'aiida.calculations:quantumespresso.xspectra':
```

File: src/aiida_quantumespresso/parsers/xspectra.py [hunk 2]
```diff
@@ -30,6 +30,8 @@ def parse(self, **kwargs):
                 return self.exit(self.exit_codes.ERROR_OUTPUT_ABSORBING_SPECIES_WRONG)
             if 'xiabs < 1 or xiabs > ntyp' in line:
                 return self.exit(self.exit_codes.ERROR_OUTPUT_ABSORBING_SPECIES_ZERO)
+            if 'Calculation not finished' in line:
+                return self.exit(self.exit_codes.ERROR_OUTPUT_TIME_LIMIT_EXCEEDED)
```
Comment (sphuber) [inline src/aiida_quantumespresso/parsers/xspectra.py:34]:
```suggestion
                return self.exit(self.exit_codes.ERROR_OUT_OF_WALLTIME)
```

File: tests/parsers/fixtures/xspectra/failed_timeout/aiida.out [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+          URL http://www.quantum-espresso.org", 
+     in publications or presentations arising from this work. More details at
+     http://www.quantum-espresso.org/quote
+
+     Parallel version (MPI), running on    16 processors
+
+     MPI processes distributed on     1 nodes
+     R & G space division:  proc/nbgrp/npool/nimage =      16
+
+     -------------------------------------------------------------------------
```
Comment (sphuber) [inline tests/parsers/fixtures/xspectra/failed_timeout/aiida.out:17]: For this test, try to cut out as much of this file as possible while the test still works. You are just testing here that if `Calculation not finished` is found at the bottom, the parser returns the right exit code.

File: tests/parsers/test_xspectra.py [hunk 4]
```diff
@@ -111,3 +111,16 @@ def test_xspectra_failed_xiabs_zero(fixture_localhost, generate_calc_job_node, g
...
+    """Test `XspectraParser` to correctly report error code 315 (time limit exceeded)."""
+    entry_point_calc_job = 'quantumespresso.xspectra'
+    entry_point_parser = 'quantumespresso.xspectra'
+
+    node = generate_calc_job_node(entry_point_calc_job, fixture_localhost, 'failed_timeout', generate_inputs())
+    parser = generate_parser(entry_point_parser)
+    _, calcfunction = parser.parse_from_node(node, store_provenance=False)
+
+    assert calcfunction.is_failed
+    assert calcfunction.exit_status == node.process_class.exit_codes.ERROR_OUTPUT_TIME_LIMIT_EXCEEDED.status
```
Comment (sphuber) [inline tests/parsers/test_xspectra.py:126]:
```suggestion
    assert calcfunction.exit_status == node.process_class.exit_codes.ERROR_OUT_OF_WALLTIME.status
```

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @PNOGillespie . Just some minor comments but overall the changes look good.

**Comment by PNOGillespie** [comment]
Thank you @sphuber for the review, I will put out a new commit after my meetings today to shorten the out-of-walltime test file.

Regarding the discussion here: https://github.com/aiidateam/aiida-quantumespresso/pull/837#discussion_r963109005 - The original formulation included FolderData as a means to provide a parent calculation for the calculation run, however the use of this wasn't tested before and I'm not familiar with the procedure for retrieving the files required in order to use the FolderData node properly for this purpose. I have opted to simply remove the option from the inputs for now and then add it back in via a later commit once I understand how this process works.

File: src/aiida_quantumespresso/calculations/xspectra.py [hunk 5]
```diff
@@ -31,7 +36,7 @@ def define(cls, spec):
 
         super().define(spec)
 
-        spec.input('parent_folder', valid_type=(orm.RemoteData, orm.FolderData), required=True)
+        spec.input('parent_folder', valid_type=(orm.RemoteData), required=True)
```
[the remark this answers is not in the document]
Comment (sphuber) [reply src/aiida_quantumespresso/calculations/xspectra.py]: Could you please still clarify this comment? Why is `FolderData` (i.e. restarting from `retrieved` folder) not supported?

File: tests/parsers/test_xspectra.py [hunk 6]
```diff
@@ -111,3 +111,16 @@ def test_xspectra_failed_xiabs_zero(fixture_localhost, generate_calc_job_node, g
 
     assert calcfunction.is_failed
     assert calcfunction.exit_status == node.process_class.exit_codes.ERROR_OUTPUT_ABSORBING_SPECIES_ZERO.status
+
+
+def test_xspectra_failed_timeout(fixture_localhost, generate_calc_job_node, generate_parser):
+    """Test `XspectraParser` to correctly report error code 315 (time limit exceeded)."""
```
Comment (sphuber) [inline tests/parsers/test_xspectra.py:117]:
```suggestion
    """Test ``XspectraParser`` to correctly report ``ERROR_OUT_OF_WALLTIME``."""
```

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @PNOGillespie . Just one minor mistake in the test docstring and ~~one outstanding comment~~. If you can address those, this can be merged.

Edit: never mind, I see you addressed that comment regarding the `FolderData` in a separate post. Just the docstring then and I will merge this.

**Comment by PNOGillespie** [comment]
Thank you for the reply @sphuber (and for spotting that mistake). I've addressed the requested change, so I'll leave the rest to you.

**Pull request merged**